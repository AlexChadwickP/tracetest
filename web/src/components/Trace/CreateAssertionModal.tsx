import {useState} from 'react';
import styled from 'styled-components';
import {Button, Input, List, Modal, Select as AntSelect, AutoComplete, Typography} from 'antd';
import jemsPath from 'jmespath';
import Text from 'antd/lib/typography/Text';

import {IAttribute, ISpan} from 'types';
import {useCreateAssertionMutation} from 'services/TestService';

const {Title} = Typography;
interface IProps {
  open: boolean;
  onClose: () => void;
  span: ISpan;
  testId: string;
  trace: any;
}

const Select = styled(AntSelect)`
  min-width: 88px;
  > .ant-select-selector {
    min-height: 100%;
  }
`;

const flattenAttributesSelector = () =>
  `resourceSpans[].[instrumentationLibrarySpans[].spans[].attributes[].{key:key,value:value.*|[0]},resource.attributes[].{key:key,value: value.*|[0]}]|[][]`;

const filterBySpanId = (spanId: string = '') =>
  `resourceSpans[?instrumentationLibrarySpans[?spans[?starts_with(spanId,'${spanId}')]]] | [].[instrumentationLibrarySpans[].spans[].attributes[].{key:key,value:value.*|[0],type:'span'},resource.attributes[].{key:key,value: value.*|[0],type:'resource'}]|[][]`;
const NORMALIZED_OBJECT = `resourceSpans[].{instrumentationLibrarySpans:instrumentationLibrarySpans[].{spans:spans[].{spanId:spanId,attributes:attributes[].{key:key,value:value.*|[0],type:'span'}}},resource:resource.{attributes:attributes[].{key:key,value: value.*|[0],type:'resource'}}}`;
const filterByAttributes = (condition: string) => `[${condition && '?'}${condition}]`;

const selectionPipe = (query: string) => `${NORMALIZED_OBJECT} | ${query}  | length([])`;
const createSelector = (query: string) => `${NORMALIZED_OBJECT} | ${query}`;
const selectorConditionBuilder = (attribute: IAttribute) => {
  if (attribute.type === 'span') {
    return `instrumentationLibrarySpans[?spans[?attributes[?key=='${attribute.key}' && value=='${attribute.value}']]]`;
  }

  if (attribute.type === 'resource') {
    return `resource.attributes[?key=='${attribute.key}' && value=='${attribute.value}']`;
  }
};

const CreateAssertionModal = ({testId, span, trace, open, onClose}: IProps) => {
  const [assertionList, setAssertionList] = useState<Array<string>>(Array(3).fill(''));
  const [createAssertion, result] = useCreateAssertionMutation();
  const attrs = jemsPath.search(trace, filterBySpanId(span.spanId));
  console.log('@@CreateAssertionModal', attrs);

  const selectorCondition = assertionList
    .map(k => {
      return attrs?.find((el: any) => el.key === k);
    })
    .filter(i => i)
    .map(item => selectorConditionBuilder(item))
    .join(' && ');
  console.log('@@query', selectionPipe(filterByAttributes(selectorCondition)));
  const effectedSpans = jemsPath.search(trace, selectionPipe(filterByAttributes(selectorCondition)));
  const spanTagsMap = attrs.reduce((acc: {[x: string]: any}, item: {key: string}) => {
    const keyPrefix = item.key.split('.').shift() || item.key;
    if (!keyPrefix) {
      return acc;
    }
    const keys = acc[keyPrefix] || [];
    keys.push(item);
    acc[keyPrefix] = keys;
    return acc;
  }, {});

  const handleAddItem = () => {
    setAssertionList([...assertionList, '']);
  };

  const renderTitle = (title: any, index: number) => (
    <span key={`KEY_${title}_${index}`}>{`${title} (${spanTagsMap[title][0].type})`}</span>
  );

  const renderItem = (attr: any) => ({
    value: attr.key,
    label: (
      <div
        style={{
          display: 'flex',
          justifyContent: 'space-between',
        }}
      >
        {attr.key}
      </div>
    ),
  });

  const keysOptions = Object.keys(spanTagsMap).map((tagKey: any, index) => {
    return {
      label: renderTitle(tagKey, index),
      options: spanTagsMap[tagKey].map((el: any) => renderItem(el)),
    };
  });

  const handleCreateAssertion = async () => {
    await createAssertion({testId, selector: createSelector(filterByAttributes(selectorCondition))});
    onClose();
  };

  return (
    <Modal style={{minHeight: 500}} footer={null} visible={span && open} onCancel={onClose}>
      <Title level={2}>Create New Assertion </Title>
      <div style={{display: 'flex', justifyContent: 'flex-end', height: 64}}>
        <Text>Effects {effectedSpans} spans</Text>
      </div>
      <List
        dataSource={assertionList}
        itemLayout="horizontal"
        loadMore={
          <Button style={{marginTop: 8}} onClick={handleAddItem}>
            Add Item
          </Button>
        }
        renderItem={(item: any, index) => {
          const handleSearchChange = (searchText: string) => {
            assertionList[index] = searchText;
            setAssertionList([...assertionList]);
          };

          return (
            <div key={`key_${index}`} style={{display: 'flex', alignItems: 'stretch', gap: 4, marginBottom: 4}}>
              <AutoComplete
                style={{width: 250}}
                onChange={handleSearchChange}
                onSearch={handleSearchChange}
                onSelect={handleSearchChange}
                options={keysOptions}
                filterOption={(inputValue, option) => {
                  return option?.label.props.children.includes(inputValue);
                }}
              >
                <Input.Search size="large" placeholder="span key" />
              </AutoComplete>
              <Select defaultValue="eq">
                <Select.Option value="eq">eq</Select.Option>
                <Select.Option value="ne">ne</Select.Option>
                <Select.Option value="gt">gt</Select.Option>
                <Select.Option value="lt">lt</Select.Option>
                <Select.Option value="ge">ge</Select.Option>
                <Select.Option value="le">le</Select.Option>
              </Select>
              <Input
                size="small"
                style={{width: 'unset'}}
                placeholder="type value"
                value={attrs?.find((el: any) => el.key === assertionList[index])?.value}
              />
            </div>
          );
        }}
      />
      <div style={{padding: 16, display: 'flex', justifyContent: 'flex-end'}}>
        <Button onClick={handleCreateAssertion}>Create</Button>
      </div>
    </Modal>
  );
};

export default CreateAssertionModal;
