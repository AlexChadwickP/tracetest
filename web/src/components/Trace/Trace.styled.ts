import {Tabs} from 'antd';
import styled from 'styled-components';

export const TraceTabs = styled(Tabs)`
  .ant-tabs-nav {
    margin-bottom: 0;
  }

  .ant-tabs-nav::before {
    border: none;
  }
`;

export const FailedTrace = styled.div`
  height: calc(100vh - 200px);
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;
