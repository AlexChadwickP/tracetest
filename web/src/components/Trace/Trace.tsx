import styled from 'styled-components';
import {ReflexContainer, ReflexSplitter, ReflexElement} from 'react-reflex';

import {Tabs} from 'antd';
import Text from 'antd/lib/typography/Text';

import 'react-reflex/styles.css';

import {useState} from 'react';

import TraceDiagram from './TraceDiagram';
import TraceTimeline from './TraceTimeline';
import TraceData from './TraceData';

import data from './data.json';
import AssertionList from './AssertionsList';
import {Test} from '../../types';

const spanMap = data.resourceSpans
  .map((i: any) => i.instrumentationLibrarySpans.map((el: any) => el.spans))
  .flat(2)
  .reduce((acc: {[key: string]: {id: string; parentIds: string[]; data: any}}, span) => {
    acc[span.spanId] = acc[span.spanId] || {id: span.spanId, parentIds: [], data: span};
    acc[span.spanId].parentIds.push(span.parentSpanId);

    return acc;
  }, {});

const Grid = styled.div`
  display: grid;
`;

const Trace = ({test}: {test: Test}) => {
  const [selectedSpan, setSelectedSpan] = useState<any>({});

  const handleSelectSpan = (span: any) => {
    setSelectedSpan(span);
  };

  return (
    <main>
      <Grid>
        <ReflexContainer style={{height: '100vh'}} orientation="horizontal">
          <ReflexElement flex={0.6}>
            <ReflexContainer orientation="vertical">
              <ReflexElement flex={0.5} className="left-pane">
                <div className="pane-content">
                  <TraceDiagram spanMap={spanMap} onSelectSpan={handleSelectSpan} selectedSpan={selectedSpan} />
                </div>
              </ReflexElement>

              <ReflexElement flex={0.5} className="right-pane">
                <div className="pane-content" style={{paddingLeft: 8}}>
                  <div>
                    <Text>Service</Text>
                  </div>
                  <Tabs>
                    <Tabs.TabPane tab="Raw Data" key="1">
                      <TraceData json={JSON.parse(JSON.stringify(selectedSpan))} />
                    </Tabs.TabPane>
                    {spanMap[selectedSpan.id]?.data && (
                      <Tabs.TabPane tab="Assertions" key="2">
                        <AssertionList testId={test.id} targetSpan={spanMap[selectedSpan.id]?.data} />
                      </Tabs.TabPane>
                    )}
                  </Tabs>
                </div>
              </ReflexElement>
            </ReflexContainer>
          </ReflexElement>
          <ReflexSplitter />
          <ReflexElement>
            <div className="pane-content">
              <TraceTimeline onSelectSpan={handleSelectSpan} selectedSpan={selectedSpan} />
            </div>
          </ReflexElement>
        </ReflexContainer>
      </Grid>
    </main>
  );
};

export default Trace;
