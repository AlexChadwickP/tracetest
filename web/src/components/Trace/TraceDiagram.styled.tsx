import styled from 'styled-components';
import {Typography} from 'antd';

enum NotchColor {
  HTTP = '#B9E28C',
  DB = '#DBCBD8',
  RPC = '#9AD4D6',
  MESSAGING = '#101935',
}

const getNotchColor = (system: string) => {
  if (system.startsWith('http.method')) {
    return NotchColor.HTTP;
  }
  if (system.startsWith('db.system')) {
    return NotchColor.DB;
  }
  if (system.startsWith('rpc.system')) {
    return NotchColor.RPC;
  }
  if (system.startsWith('messaging.system')) {
    return NotchColor.MESSAGING;
  }
  return '#F49D6E';
};

export const TraceNode = styled.div<{selected: boolean}>`
  background-color: white;
  border: 2px solid ${({selected}) => (selected ? 'rgb(0, 161, 253)' : 'rgb(213, 215, 224)')};
  border-radius: 16px;
  min-width: fit-content;
  display: flex;
  width: 200px;
  height: 80px;
  justify-content: center;
  align-items: center;
`;

export const TraceNotch = styled.div<{system: string}>`
  background-color: ${({system}) => getNotchColor(system)};
  margin-top: -8px;
  position: absolute;
  top: 0px;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 16px;
  padding-right: 16px;
  border-radius: 16px;
  width: 70%;
  justify-content: center;
  align-items: center;
  text-align: center;
`;

export const Container = styled.div`
  position: relative;
`;

export const LoadingLabel = styled(Typography.Text)`
  position: absolute;
  top: 30%;
  left: 50%;
  transform: translate(-50%, -50%);
  color: #c4c4c4;
  font-size: 24px;
  line-height: 32px;
  font-weight: 600;
`;
