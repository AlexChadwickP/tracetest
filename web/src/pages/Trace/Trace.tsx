import styled from 'styled-components';
import {ReflexContainer, ReflexSplitter, ReflexElement} from 'react-reflex';
import Title from 'antd/lib/typography/Title';

import 'react-reflex/styles.css';

import {useState} from 'react';

import TraceDiagram from './TraceDiagram';
import TraceTimeline from './TraceTimeline';
import TraceData from './TraceData';

import data from './data.json';

const spanMap = data.data
  .map(i => i.spans)
  .flat()
  .reduce((acc: {[key: string]: {id: string; parentIds: string[]; data: any}}, span) => {
    acc[span.spanID] = acc[span.spanID] || {id: span.spanID, parentIds: [], data: span};
    span.references.forEach(p => {
      acc[span.spanID].parentIds.push(p.spanID);
    });
    return acc;
  }, {});

const Grid = styled.div`
  display: grid;
`;

const Header = styled.div`
  display: flex;
  align-items: center;
  width: 100%;
  height: 64px;
  padding: 0 32px;
  border-bottom: 1px solid rgb(213, 215, 224);
`;

const Trace = () => {
  const [selectedSpan, setSelectedSpan] = useState<any>({});

  const handleSelectSpan = (span: any) => {
    setSelectedSpan(span);
  };

  return (
    <main>
      <Header>
        <Title level={3}>{data.data[0].spans[0].operationName}</Title>
      </Header>
      <Grid>
        <ReflexContainer style={{minHeight: 1000, height: '100%'}} orientation="horizontal">
          <ReflexElement size={500} maxSize={500}>
            <ReflexContainer style={{height: '100%'}} orientation="vertical">
              <ReflexElement flex={0.5} className="left-pane">
                <div className="pane-content">
                  <TraceDiagram spanMap={spanMap} onSelectSpan={handleSelectSpan} selectedSpan={selectedSpan} />
                </div>
              </ReflexElement>
              <ReflexSplitter />
              <ReflexElement flex={0.5} className="right-pane">
                <div className="pane-content">
                  <TraceData json={JSON.parse(JSON.stringify(selectedSpan))} />
                </div>
              </ReflexElement>
            </ReflexContainer>
          </ReflexElement>
          <ReflexSplitter />
          <ReflexElement>
            <div className="pane-content">
              <TraceTimeline onSelectSpan={handleSelectSpan} selectedSpan={selectedSpan} />
            </div>
          </ReflexElement>
        </ReflexContainer>
      </Grid>
    </main>
  );
};

export default Trace;
