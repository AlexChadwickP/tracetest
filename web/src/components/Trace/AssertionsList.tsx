import React, {useState} from 'react';
import {Button} from 'antd';

import jemsPath from 'jmespath';

import {filterBySpanId} from 'utils';
import {IAttribute, ISpan} from 'types';
import {useGetTestAssertionsQuery} from 'services/TestService';
import AssertionsResultTable from 'components/AssertionsTable/AssertionsTable';

import CreateAssertionModal from './CreateAssertionModal';

interface IProps {
  testId: string;
  targetSpan: ISpan;
  trace: any;
}

const AssertionList = ({testId, targetSpan, trace}: IProps) => {
  const [openCreateAssertion, setOpenCreateAssertion] = useState(false);
  const {data: testAssertions} = useGetTestAssertionsQuery(testId);
  const attrs: IAttribute[] = jemsPath.search(trace, filterBySpanId(targetSpan.spanId));

  const attributesTree = attrs?.reduce((acc: any, item: any) => {
    const resource = acc[item.type] || {};
    resource.title = item.type;
    resource.key = item.type;
    resource.children = resource.children || [];
    resource.children.push({title: `${item.key} = ${item.value}`, key: item.key});
    acc[item.type] = resource;
    return acc;
  }, {});

  return (
    <div>
      <Button style={{marginBottom: 8}} onClick={() => setOpenCreateAssertion(true)}>
        New Assertion
      </Button>
      <AssertionsResultTable />
      <CreateAssertionModal
        key={`KEY_${targetSpan.spanId}`}
        testId={testId}
        trace={trace}
        span={targetSpan}
        open={openCreateAssertion}
        onClose={() => setOpenCreateAssertion(false)}
      />
    </div>
  );
};

export default AssertionList;
