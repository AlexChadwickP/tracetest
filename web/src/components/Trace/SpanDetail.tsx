import {Button, Typography} from 'antd';
import {skipToken} from '@reduxjs/toolkit/dist/query';
import {PlusOutlined} from '@ant-design/icons';
import {FC, useMemo, useState} from 'react';
import {useGetTestByIdQuery} from 'services/TestService';
import {Assertion, ISpan, ITrace, SpanAssertionResult} from 'types';
import {runAssertionBySpanId} from 'services/AssertionService';
import AssertionsResultTable from 'components/AssertionsTable/AssertionsTable';
import SkeletonTable from 'components/SkeletonTable';
import * as S from './SpanDetail.styled';
import CreateAssertionModal from './CreateAssertionModal';
import Attributes from './Attributes';

type TSpanDetailProps = {
  testId: string;
  targetSpan?: ISpan;
  trace?: ITrace;
};

const SpanDetail: FC<TSpanDetailProps> = ({testId, targetSpan, trace}) => {
  const [openCreateAssertion, setOpenCreateAssertion] = useState(false);
  const {data: test} = useGetTestByIdQuery(testId ?? skipToken);

  const assertionsResultList = useMemo(() => {
    if (!targetSpan?.spanId || !trace) {
      return [];
    }
    return (
      test?.assertions?.reduce<Array<{assertion: Assertion; assertionResultList: Array<SpanAssertionResult>}>>(
        (resultList, assertion) => {
          const assertionResultList = runAssertionBySpanId(targetSpan.spanId, trace, assertion);

          return assertionResultList ? [...resultList, {assertion, assertionResultList}] : resultList;
        },
        []
      ) || []
    );
  }, [targetSpan?.spanId, test?.assertions, trace]);

  return (
    <>
      <S.DetailsContainer>
        <S.DetailsHeader>
          <Typography.Text strong> Details for selected span</Typography.Text>
          <Button type="link" icon={<PlusOutlined />} onClick={() => setOpenCreateAssertion(true)}>
            Add Assertion
          </Button>
        </S.DetailsHeader>
        <SkeletonTable loading={!targetSpan || !trace}>
          {assertionsResultList.length ? (
            assertionsResultList.map(({assertion, assertionResultList}, index) => (
              <AssertionsResultTable
                key={assertion.assertionId}
                assertionResults={assertionResultList}
                sort={index + 1}
                assertion={assertion}
              />
            ))
          ) : (
            <S.DetailsEmptyStateContainer>
              <S.DetailsTableEmptyStateIcon />
              <Typography.Text disabled>No Data</Typography.Text>
            </S.DetailsEmptyStateContainer>
          )}
        </SkeletonTable>
      </S.DetailsContainer>
      <Attributes spanId={targetSpan?.spanId} trace={trace} />

      {targetSpan?.spanId && (
        <CreateAssertionModal
          key={`KEY_${targetSpan?.spanId}`}
          testId={testId}
          trace={trace}
          span={targetSpan!}
          open={openCreateAssertion}
          onClose={() => setOpenCreateAssertion(false)}
        />
      )}
    </>
  );
};

export default SpanDetail;
