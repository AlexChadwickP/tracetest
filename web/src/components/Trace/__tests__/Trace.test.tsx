import {render} from '@testing-library/react';
import {ReactFlowProvider} from 'react-flow-renderer';
import {ReduxWrapperProvider} from '../../../redux/ReduxWrapperProvider';
import {TestingModels} from '../../../utils/TestingModels';
import Trace from '../Trace';

test('Trace', () => {
  const result = render(
    <ReactFlowProvider>
      <ReduxWrapperProvider>
        <div style={{width: 600, height: 600}}>
          <Trace
            minHeight="300px"
            testResultDetails={TestingModels.testRunResult}
            test={TestingModels.test}
            visiblePortion={100}
            displayError={false}
          />
        </div>
      </ReduxWrapperProvider>
    </ReactFlowProvider>
  );
  expect(result.container).toMatchSnapshot();
});
