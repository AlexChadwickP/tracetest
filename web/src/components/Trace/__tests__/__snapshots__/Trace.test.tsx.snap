// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trace 1`] = `
<div>
  <div
    style="width: 600px; height: 600px;"
  >
    <main
      class="sc-dJjYzT iVWlrj"
      height="300px"
    >
      <div
        class="sc-hGPBjI cVQbfo"
      >
        <div
          class="sc-fKVqWL bfwmLL"
        >
          <div
            class="sc-bBHxTw eTRexs"
          >
            <span
              aria-label="cluster"
              class="anticon anticon-cluster sc-iwjdpV grVIcu"
              role="img"
              style="font-size: 18px;"
              tabindex="-1"
            >
              <svg
                aria-hidden="true"
                data-icon="cluster"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M888 680h-54V540H546v-92h238c8.8 0 16-7.2 16-16V168c0-8.8-7.2-16-16-16H240c-8.8 0-16 7.2-16 16v264c0 8.8 7.2 16 16 16h238v92H190v140h-54c-4.4 0-8 3.6-8 8v176c0 4.4 3.6 8 8 8h176c4.4 0 8-3.6 8-8V688c0-4.4-3.6-8-8-8h-54v-72h220v72h-54c-4.4 0-8 3.6-8 8v176c0 4.4 3.6 8 8 8h176c4.4 0 8-3.6 8-8V688c0-4.4-3.6-8-8-8h-54v-72h220v72h-54c-4.4 0-8 3.6-8 8v176c0 4.4 3.6 8 8 8h176c4.4 0 8-3.6 8-8V688c0-4.4-3.6-8-8-8zM256 805.3c0 1.5-1.2 2.7-2.7 2.7h-58.7c-1.5 0-2.7-1.2-2.7-2.7v-58.7c0-1.5 1.2-2.7 2.7-2.7h58.7c1.5 0 2.7 1.2 2.7 2.7v58.7zm288 0c0 1.5-1.2 2.7-2.7 2.7h-58.7c-1.5 0-2.7-1.2-2.7-2.7v-58.7c0-1.5 1.2-2.7 2.7-2.7h58.7c1.5 0 2.7 1.2 2.7 2.7v58.7zM288 384V216h448v168H288zm544 421.3c0 1.5-1.2 2.7-2.7 2.7h-58.7c-1.5 0-2.7-1.2-2.7-2.7v-58.7c0-1.5 1.2-2.7 2.7-2.7h58.7c1.5 0 2.7 1.2 2.7 2.7v58.7zM360 300a40 40 0 1080 0 40 40 0 10-80 0z"
                />
              </svg>
            </span>
            <span
              aria-label="bars"
              class="anticon anticon-bars sc-cxpSdN gHBSOa"
              role="img"
              style="font-size: 18px;"
              tabindex="-1"
            >
              <svg
                aria-hidden="true"
                data-icon="bars"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="0 0 1024 1024"
                width="1em"
              >
                <path
                  d="M912 192H328c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 284H328c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 284H328c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM104 228a56 56 0 10112 0 56 56 0 10-112 0zm0 284a56 56 0 10112 0 56 56 0 10-112 0zm0 284a56 56 0 10112 0 56 56 0 10-112 0z"
                />
              </svg>
            </span>
          </div>
          <span
            class="ant-input-affix-wrapper sc-llYSUQ iKhxFO"
          >
            <span
              class="ant-input-prefix"
            >
              <span
                aria-label="search"
                class="anticon anticon-search"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="search"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                  />
                </svg>
              </span>
            </span>
            <input
              class="ant-input"
              placeholder="Search in trace"
              type="text"
              value=""
            />
          </span>
        </div>
        <div
          class="sc-eCImPb kQLCbn"
          data-cy="diagram-dag"
        >
          <div
            class="react-flow"
          >
            <div
              class="react-flow__renderer react-flow__zoompane"
            >
              <div
                class="react-flow__nodes"
                style="transform: translate(0px,0px) scale(0.5);"
              >
                <div
                  class="react-flow__node react-flow__node-TraceNode selected selected selectable"
                  data-id="4938u43"
                  style="z-index: 10; transform: translate(100px,75px); pointer-events: all; opacity: 0;"
                >
                  <div
                    class="sc-bdvvtL fGMamS"
                    data-cy="trace-node-http"
                  >
                    <div
                      class="sc-gsDKAQ cmaxSH"
                    >
                      <span
                        class="ant-typography"
                      >
                        HTTP
                      </span>
                    </div>
                    <div
                      class="react-flow__handle react-flow__handle-top nodrag target connectable"
                      data-handleid="4938u43"
                      data-handlepos="top"
                      data-nodeid="4938u43"
                      style="top: 0px; border-radius: 0; visibility: hidden;"
                    />
                    <div
                      class="sc-hKwDye etMbsh"
                    >
                      
                      <span
                        class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line sc-dkPtRN dvhzfk"
                      >
                        
                      </span>
                    </div>
                    <div
                      class="react-flow__handle react-flow__handle-bottom nodrag source connectable"
                      data-handleid="4938u43"
                      data-handlepos="bottom"
                      data-nodeid="4938u43"
                      style="bottom: 0px; border-radius: 0; visibility: hidden;"
                    />
                  </div>
                </div>
              </div>
              <svg
                class="react-flow__edges"
                height="500"
                width="500"
              >
                <defs>
                  <marker
                    class="react-flow__arrowhead"
                    id="react-flow__arrowclosed"
                    markerHeight="12.5"
                    markerWidth="12.5"
                    orient="auto"
                    refX="0"
                    refY="0"
                    viewBox="-10 -10 20 20"
                  >
                    <polyline
                      fill="#b1b1b7"
                      points="-5,-4 0,0 -5,4 -5,-4"
                      stroke="#b1b1b7"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="1"
                    />
                  </marker>
                  <marker
                    class="react-flow__arrowhead"
                    id="react-flow__arrow"
                    markerHeight="12.5"
                    markerWidth="12.5"
                    orient="auto"
                    refX="0"
                    refY="0"
                    viewBox="-10 -10 20 20"
                  >
                    <polyline
                      fill="none"
                      points="-5,-4 0,0 -5,4"
                      stroke="#b1b1b7"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="1.5"
                    />
                  </marker>
                </defs>
                <g
                  style="transform: translate(0px,0px) scale(0.5);"
                />
              </svg>
              <div
                class="react-flow__pane"
              />
            </div>
            <svg
              class="react-flow__background"
              style="width: 100%; height: 100%;"
            >
              <pattern
                height="2"
                id="pattern-40110"
                patternUnits="userSpaceOnUse"
                width="2"
                x="0"
                y="0"
              >
                <circle
                  cx="0.5"
                  cy="0.5"
                  fill="#FBFBFF"
                  r="0.5"
                />
              </pattern>
              <rect
                fill="url(#pattern-40110)"
                height="100%"
                width="100%"
                x="0"
                y="0"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="sc-dlVxhl hKPJke"
      >
        <div
          class="sc-fotOHu gyNzhm"
        >
          <div
            class="sc-iqseJM gLNCmO"
          >
            <h5
              class="ant-typography sc-crHmcD fbxtEG"
            >
              Span Details
            </h5>
            <span
              class="ant-typography"
            >
              HTTP •  • 
            </span>
          </div>
          <div
            class="ant-tabs ant-tabs-top sc-egiyK grcuFy"
            data-cy="span-details-attributes"
            style="margin-bottom: 150px; overflow: hidden;"
          >
            <div
              class="ant-tabs-nav"
              role="tablist"
            >
              <div
                class="ant-tabs-nav-wrap"
              >
                <div
                  class="ant-tabs-nav-list"
                  style="transform: translate(0px, 0px);"
                >
                  <div
                    class="ant-tabs-tab ant-tabs-tab-active"
                  >
                    <div
                      aria-controls="rc-tabs-test-panel-span-assertion"
                      aria-selected="true"
                      class="ant-tabs-tab-btn"
                      id="rc-tabs-test-tab-span-assertion"
                      role="tab"
                      tabindex="0"
                    >
                      Assertion
                    </div>
                  </div>
                  <div
                    class="ant-tabs-tab"
                  >
                    <div
                      aria-controls="rc-tabs-test-panel-span-request"
                      aria-selected="false"
                      class="ant-tabs-tab-btn"
                      id="rc-tabs-test-tab-span-request"
                      role="tab"
                      tabindex="0"
                    >
                      Request
                    </div>
                  </div>
                  <div
                    class="ant-tabs-tab"
                  >
                    <div
                      aria-controls="rc-tabs-test-panel-span-response"
                      aria-selected="false"
                      class="ant-tabs-tab-btn"
                      id="rc-tabs-test-tab-span-response"
                      role="tab"
                      tabindex="0"
                    >
                      Response
                    </div>
                  </div>
                  <div
                    class="ant-tabs-tab"
                  >
                    <div
                      aria-controls="rc-tabs-test-panel-span-attribute-list"
                      aria-selected="false"
                      class="ant-tabs-tab-btn"
                      id="rc-tabs-test-tab-span-attribute-list"
                      role="tab"
                      tabindex="0"
                    >
                      Attribute list
                    </div>
                  </div>
                  <div
                    class="ant-tabs-ink-bar ant-tabs-ink-bar-animated"
                  />
                </div>
              </div>
              <div
                class="ant-tabs-nav-operations ant-tabs-nav-operations-hidden"
              >
                <button
                  aria-controls="rc-tabs-test-more-popup"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-hidden="true"
                  class="ant-tabs-nav-more"
                  id="rc-tabs-test-more"
                  style="visibility: hidden; order: 1;"
                  tabindex="-1"
                  type="button"
                >
                  <span
                    aria-label="ellipsis"
                    class="anticon anticon-ellipsis"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="ellipsis"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
                      />
                    </svg>
                  </span>
                </button>
              </div>
            </div>
            <div
              class="ant-tabs-content-holder"
            >
              <div
                class="ant-tabs-content ant-tabs-content-top"
              >
                <div
                  aria-hidden="false"
                  aria-labelledby="rc-tabs-test-tab-span-assertion"
                  class="ant-tabs-tabpane ant-tabs-tabpane-active"
                  id="rc-tabs-test-panel-span-assertion"
                  role="tabpanel"
                  tabindex="0"
                >
                  <div>
                    <div
                      class="sc-fFeiMQ MSqGW"
                    >
                      <button
                        class="ant-btn ant-btn-primary ant-btn-background-ghost sc-bkkeKt djgNJH"
                        data-cy="add-assertion-button"
                        type="button"
                      >
                        <span
                          aria-label="plus"
                          class="anticon anticon-plus"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="plus"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <defs>
                              <style />
                            </defs>
                            <path
                              d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
                            />
                            <path
                              d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
                            />
                          </svg>
                        </span>
                        <span>
                          Add Assertion
                        </span>
                      </button>
                    </div>
                    <div
                      class="sc-ksdxgE eejQJE"
                      data-cy="empty-assertion-table"
                    >
                      <img
                        class="sc-hBUSln"
                        src="SpanAssertionsEmptyState.svg"
                      />
                      <span
                        class="ant-typography ant-typography-disabled"
                      >
                        No Data
                      </span>
                    </div>
                  </div>
                </div>
                <div
                  aria-hidden="true"
                  aria-labelledby="rc-tabs-test-tab-span-request"
                  class="ant-tabs-tabpane"
                  id="rc-tabs-test-panel-span-request"
                  role="tabpanel"
                  style="display: none;"
                  tabindex="-1"
                />
                <div
                  aria-hidden="true"
                  aria-labelledby="rc-tabs-test-tab-span-response"
                  class="ant-tabs-tabpane"
                  id="rc-tabs-test-panel-span-response"
                  role="tabpanel"
                  style="display: none;"
                  tabindex="-1"
                />
                <div
                  aria-hidden="true"
                  aria-labelledby="rc-tabs-test-tab-span-attribute-list"
                  class="ant-tabs-tabpane"
                  id="rc-tabs-test-panel-span-attribute-list"
                  role="tabpanel"
                  style="display: none;"
                  tabindex="-1"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </main>
  </div>
</div>
`;
